### PR TITLE
Hyperelasticity example

### PR DESCRIPTION
**Pull request by bhaveshshrimali**
This PR adds an example illustrating a mixed finite element implementation of (nearly) incompressible hyperelasticity #455 using the Taylor-Hood element. It also adds two helper functions `inv` and `det` to `skfem.helpers.py` to compute the inverse and determinant of higher order tensors over trailing axes.

### CODE REVIEW

**Comment by bhaveshshrimali** [comment]
trialFigureWhiteBG

A tiny bit more discernible?

**Comment by gdmcbain** [comment]
Is the solution supposed to be isobaric?  If so, plotting the pressure seems redundant.

I think the local magnitude of the velocity is more often written |_u_|, reserving ‖_u_‖ for a norm of the velocity field.

**Comment by bhaveshshrimali** [comment]
> Is the solution supposed to be isobaric? If so, plotting the pressure seems redundant.

The pressure for a homogeneous domain under uniaxial loading should be equal to `-mu/l`, which is exactly recovered by the `P1` elements in this case. But I agree the plot doesn't reveal any more insight than this.

**Comment by bhaveshshrimali** [comment]
> I think the local magnitude of the velocity is more often written |_u_|, reserving ‖_u_‖ for a norm of the velocity field.

Ah, that was a bit sloppy. I will change this now. Thanks!!

As an aside, do you know the reason behind the failing build? I saw `pacopy` was pinned to `0.1.2` due to the licensing issue recently.

**Comment by gdmcbain** [comment]
> As an aside, do you know the reason behind the failing build? I saw pacopy was pinned to 0.1.2 due to the licensing issue recently.

Hmm.  I'll look into it.  I think the pacopy stuff was all passing when #452 #453 #457 #459 #461 were merged, but let me check.

An aside to aside: I'm doing much work on re-implementing natural parameter continuation without pacopy, as the latter gets in the way of JFNK #119.  I expect that some kind of continuation (natural parameter or perhaps pseudo-transient) will be helpful in some more nonlinear problems in elasticity.

**Comment by bhaveshshrimali** [comment]
> expect that some kind of continuation (natural parameter or perhaps pseudo-transient) will be helpful in some more nonlinear problems in elasticity.

Yeah some cool nonlinear buckling problems that often arise in rubber-like hyperelastic materials . I haven't gone to solving those yet either so definitely a nice problem to try out in the future. Thanks

**Comment by gdmcbain** [comment]
I see, yes: https://travis-ci.com/github/kinnala/scikit-fem/jobs/376714751 has

> ERROR: Could not find a version that satisfies the requirement pacopy==0.1.2 (from versions: none)

> ERROR: No matching distribution found for pacopy==0.1.2

> The command "pip install meshio pacopy==0.1.2 pygmsh pyamg flake8" failed and exited with 1 during .

I'll raise an issue for this: #463.

**Comment by gdmcbain** [comment]
In looking into the Travis CI failure #463, it so happened that I saw something else that needed fixing (unnecessary apt install of libsuitesparse #464), so I submitted a PR to omit that #465 and it passed.  I'm pretty sure that (_a_) your new example here is independent of pacopy and (_b_) #465 is independent of pacopy, so I'm puzzled that #465 passes but this PR fails.

**Comment by bhaveshshrimali** [comment]
> ERROR: Could not find a version that satisfies the requirement pacopy==0.1.2 (from versions: none)

I think this is no longer occuring, but I'm confused as to why it was before. I am addressing the whitespace/newline issues.

**Comment by bhaveshshrimali** [comment]
trialFigureWhiteBGdispOnly

Removed the pressure contour and edited the legend for the displacement plot. I will resume this later tomorrow to check for typos etc. in the description, but feel free to suggest changes. 

I have tried to keep the description short and yet not excessively cut on essential details that would describe the problem sufficiently well.

File: skfem/helpers.py [hunk 1]
```diff
@@ -97,3 +97,63 @@ def eye(w, n):
     """Create diagonal matrix with w on diagonal."""
     return np.array([[w if i == j else 0. * w for i in range(n)]
                      for j in range(n)])
+
+
+def det(A):
```
Comment (kinnala) [inline skfem/helpers.py:102]: Should be easy to make it work also for 2-by-2 matrices? `if A.shape[0]==2: ...`

File: skfem/helpers.py [hunk 2]
```diff
@@ -97,3 +97,63 @@ def eye(w, n):
...
+    ----------
+    A : (N, N,...) numpy.ndarray
+        Input array whose inverse is to be computed
+    Returns
+    -------
+    Ainv : (N, N,...) numpy.ndarray
+        Inverse of `A`.
+    """
+    invA = zeros_like(A)
+    detA = A[0, 0] * (A[1, 1] * A[2, 2] - A[1, 2] * A[2, 1]) -\
```
Comment (kinnala) [inline skfem/helpers.py:137]: Can this use `detA = det(A)` ?

File: skfem/helpers.py [hunk 3]
```diff
@@ -97,3 +97,63 @@ def eye(w, n):
...
+    detA = A[0, 0] * (A[1, 1] * A[2, 2] -
+                      A[1, 2] * A[2, 1]) -\
+        A[0, 1] * (A[1, 0] * A[2, 2] -
+                   A[1, 2] * A[2, 0]) +\
+        A[0, 2] * (A[1, 0] * A[2, 1] -
+                   A[1, 1] * A[2, 0])
+    return detA
+
+
+def inv(A):
```
Comment (kinnala) [inline skfem/helpers.py:124]: It should also be easy to make this work for 2x2 matrices.

File: docs/examples/ex36.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
+r"""Incompressible Hyperelasticity
+This example solves the governing equations describing the mechanical
```
Comment (kinnala) [inline docs/examples/ex36.py:2]: Can you also add a short version of the description to `docs/listofexamples.rst` together with the image?

The image can be uploaded to this issue https://github.com/kinnala/scikit-fem/issues/431 after which you can get an url with right click.

**Review by kinnala — CHANGES_REQUESTED** [review]
Some improvements.

File: skfem/helpers.py (hunk 3 above)
Comment (bhaveshshrimali) [reply skfem/helpers.py:124]: Ah I see, completely forgot that these were to be modified to include the `2x2` cases as well.

File: skfem/helpers.py (hunk 1 above)
Comment (bhaveshshrimali) [reply skfem/helpers.py:102]: I made these changes. Would test a couple more things. For instance, now that I remember it should be possible to write the inverse entirely using detA and `trace(A^2)` and `trace(A` and implementing these using `np.einsum`. If this is performant then would be valid for matrices of any rank (although I don't know where would these cases arise)

Same goes for `det(A)`.

Comment (bhaveshshrimali) [reply skfem/helpers.py:102]: Just tested the Cayley-Hamilton version

```python3
def finv(A):
    detA = vdet(A)
    I1 = einsum("ii...", A)
    I2 = -0.5 * (einsum("ik...,ki...", A, A) - I1**2)
    Asq = einsum("ik...,kj...->ij...", A, A)
    eye = zeros_like(A)
    eye[0, 0] = 1.
    eye[1, 1] = 1.
    eye[2,2] = 1.
    return 1./detA * (Asq - I1 * A + I2 * eye)
```

It's significantly (`~10x`) slower and more complicated than your version. Still faster than `numpy.linalg.inv` on my machine, which for some reason is extremely slow

**Comment by kinnala** [comment]
Fantastic. I think this is ready. I'll merge it later today and check that the documentation recompiles properly.

**Comment by bhaveshshrimali** [comment]
Thanks a lot for your help. I did open up an [issue on numpy](https://github.com/numpy/numpy/issues/17166) regarding `inv` being very slow for such cases :)

**Comment by kinnala** [comment]
The figure was created with vedo?

**Comment by bhaveshshrimali** [comment]
No, Paraview.

**Comment by kinnala** [comment]
Thanks. I made some small code formatting changes and also decided to remove the saddle point problem from the short description because I didn't feel like the notation was familiar to non-experts. Better to read the entire description from ex36.py.

Thanks a lot for your contribution to scikit-fem. You are now entitled to have your name and ORCiD added to the Zenodo archive: https://github.com/kinnala/scikit-fem/blob/master/.zenodo.json If you wish to be included, please open another PR that adds your information.

**Comment by bhaveshshrimali** [comment]
> If you wish to be included, please open another PR that adds your information.

Thanks a lot for considering it. I will open a PR for the same.

**Pull request merged**